File: src/main/java/org/dedda/bratwurst/parse/BWClassParser.java
package org.dedda.bratwurst.parse;

import org.dedda.bratwurst.lang.BWClass;
import org.dedda.bratwurst.lang.BWFunction;

import java.util.LinkedList;
import java.util.List;

import static org.dedda.bratwurst.parse.Patterns.*;

/**
 * Created by dedda on 9/25/15.
 *
 * @author dedda
 */
public class BWClassParser {

    public BWClass parse(String[] lines, int begin) {
        int end = getEnd(lines, begin);
        for (int i = 0; i < lines.length; i++) {
            lines[i] = lines[i].trim();
        }
        String className = null;
        List<BWFunction> functions = new LinkedList<>();
        BWFunctionParser functionParser = new BWFunctionParser();
        for (int i = begin; i < end; i++) {
            String line = lines[i];
            if (line.matches(NAMING)) {
                className = line.split(" ")[2];
                continue;
            }
            if (line.matches(FUNCTION_BEGIN)) {
                functions.add(functionParser.parse(lines, i));
                i = functionParser.getEndOfFunction(lines, i);
            }
        }
        if (className == null) {
            throw new RuntimeException("class name not found!");
        }
        BWFunction[] functionsArray = new BWFunction[functions.size()];
        functions.toArray(functionsArray);
        return new BWClass(className, functionsArray);
    }

    public int getEnd(String[] lines, int begin) {
        for (int i = begin; i < lines.length; i++) {
            if (lines[i].trim().matches(CLASS_END)) {
                return i;
            }
        }
        throw new RuntimeException("no class end found!");
    }

}

File: src/main/java/org/dedda/bratwurst/parse/BWClassParser.kt
package org.dedda.bratwurst.parse

import org.dedda.bratwurst.lang.BWFunction
import org.dedda.bratwurst.lang.classes.BWClass
import java.util.*
import java.util.stream.IntStream

/**
 * Created by dedda on 9/25/15.
 *
 * @author dedda
 */
class BWClassParser {
    fun parse(lines: List<String>, begin: Int): BWClass {
        val end = getEnd(lines, begin)
        val trimmed = lines.map(String::trim)

        var className: String? = null
        val functions: MutableList<BWFunction> = LinkedList()
        val functionParser = BWFunctionParser()
        var i = begin
        while (i < end) {
            val line = trimmed[i]
            if (line.matches(Regex(Patterns.NAMING))) {
                className = line.split(" ").toTypedArray()[2]
                i++
                continue
            }
            if (line.matches(Regex(Patterns.FUNCTION_BEGIN))) {
                functions.add(functionParser.parse(trimmed, i))
                i = functionParser.getEndOfFunction(trimmed, i)
            }
            i++
        }
        if (className == null) {
            throw RuntimeException("class name not found!")
        }
        val functionsArray = functions.toTypedArray()
        return BWClass(className, functionsArray)
    }

    fun getEnd(lines: List<String>, begin: Int): Int {
        val pattern = Regex(Patterns.CLASS_END)
        return IntStream.range(begin, lines.size)
            .filter { line -> lines[line].trim().matches(pattern) }
            .findFirst()
            .orElseThrow {
                throw RuntimeException("no class end found!")
            }
    }
}